### PR TITLE
tag result pages

### LINKED ISSUE #10789: Tag filter is not visible in the search results
The list of search results can be filtered by tag, but this is not apparent from the UI.

For example, view the list of add-ons tagged `privacy` by visiting:
https://addons.allizom.org/en-US/firefox/search/?tag=privacy
Search for something else, e.g. "privacy". The URL becomes
https://addons.allizom.org/en-US/firefox/search/?q=privacy&tag=privacy

Both pages are similar, except for the list of results. The visibility of the tag filter should be improved, e.g. by changing `N results found for "(query)"` to `N results found for "(query)" filtered by "(tag, optionally with link to full list of add-ons by tag)", and/or by showing the tag filter(s) in the "Filter results" section at the left.

### PR DESCRIPTION
**Pull request by eviljeff**
fixes #10789

### CODE REVIEW

**Comment by eviljeff** [comment]
Seems to work (might be missing some edge cases) - it's based on category pages but it seemed like it would be a lot more complicated to have the same code handle both types (category pages needing a more complicated i18n solution ).  

1 failing test I don't entirely understand the cause for (works for category pages!).

File: src/amo/components/SearchContextCard/index.js [hunk 1]
```diff
@@ -164,6 +164,12 @@ export class SearchContextCardBase extends React.Component<InternalProps> {
     } else if (loadingSearch) {
       searchText = i18n.gettext('Searching for add-ons');
     }
+    if (searchText && tag) {
+      const tagText = i18n.sprintf(i18n.gettext('with tag %(tag)s'), {
+        tag,
+      });
+      searchText = `${searchText} ${tagText}`;
+    }
```
Comment (willdurand) [inline src/amo/components/SearchContextCard/index.js:172]: We need full translated strings for localizers so we cannot do that unfortunately.

Comment (willdurand) [reply src/amo/components/SearchContextCard/index.js:172]: And I guess we only want the translation for tags when we do have search results and not when it's loading.

File: src/amo/pages/TagPage/index.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+};
+
+export const extractId = (ownProps: Props): string => {
+  return ownProps.match.params.tag;
+};
+
+const TagPage: React.ComponentType<Props> = compose(
+  connect(mapStateToProps),
+  translate(),
+  withFixedErrorHandler({ fileName: __filename, extractId }),
```
Comment (willdurand) [inline src/amo/pages/TagPage/index.js:90]: Do we really need an error handler here? It does not seem used anywhere.

File: src/amo/pages/TagPage/index.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    });
+  }
+
+  render(): React.Node {
+    const { filters, match } = this.props;
+    const { tag } = match.params;
+
+    const filtersForSearch = {
+      ...filters,
+      tag,
```
Comment (willdurand) [inline src/amo/pages/TagPage/index.js:54]: Please add a comment to indicate that `filters` does not contain a `tag` because it is deleted in `src/amo/components/SearchFilters/index.js`.

File: src/amo/components/SearchFilters/index.js [hunk 4]
```diff
@@ -128,6 +128,9 @@ export class SearchFiltersBase extends React.Component<InternalProps> {
       delete filters.category;
       delete filters.addonType;
     }
+    if (filters.tag) {
```
Comment (willdurand) [inline src/amo/components/SearchFilters/index.js:131]: Please add a comment similar to the one for category. This isn't something we normally do (deleting filters).

**Review by willdurand — COMMENTED** [review]
Great start! Here is an early round of comments.

**Comment by willdurand** [comment]
> Seems to work (might be missing some edge cases) - it's based on category pages but it seemed like it would be a lot more complicated to have the same code handle both types (category pages needing a more complicated i18n solution ).

Sounds good to me.

> 1 failing test I don't entirely understand the cause for (works for category pages!).

I think that's because of this: https://github.com/mozilla/addons-frontend/blob/9d4cdb7a26fb653de2f3bd22ce4d58947fab02a3/src/amo/components/SearchFilters/index.js#L73-L85

If you want to fully mimic the category behavior (which seems to be implied by the constant `DEFAULT_TAG_SORT` introduced in this PR), you should probably update `onSelectElementChange()` too.

This isn't a very good test because it is kinda covering two behaviors at once and the description only mentions one of the two behaviors... Ideally, we'd have one test case for the filter being removed and another one for the `sort` value being updated to include `recommended` when changed.

**Comment by eviljeff** [comment]
> > 1 failing test I don't entirely understand the cause for (works for category pages!).
> 
> I think that's because of this:
> 
> https://github.com/mozilla/addons-frontend/blob/9d4cdb7a26fb653de2f3bd22ce4d58947fab02a3/src/amo/components/SearchFilters/index.js#L73-L85
> 
> If you want to fully mimic the category behavior (which seems to be implied by the constant `DEFAULT_TAG_SORT` introduced in this PR), you should probably update `onSelectElementChange()` too.

added in https://github.com/mozilla/addons-frontend/pull/10820/commits/081fda0540b112e916b4f6feb16d3be747122e5f along with the comments.  

I've realised there's a limitation in just always dropping `tag` in the `SearchFilters` component - if there's a `tag=xxx` param in a vanilla search page you also lose it by changing the filters (e.g. change the sort) - not just actions like searching in the search box.  Really, we want to retain `tag` when the UI filters are changed and only drop it when there's a new search.

We don't have this problem with searching with categories because we redirect search queries involving categories to the category pages.  (Something we weren't planning to do with tag pages - and if we did we'd need a different url because of the nginx 301).

File: src/amo/components/SearchContextCard/index.js (hunk 1 above)
Comment (eviljeff) [reply src/amo/components/SearchContextCard/index.js:172]: 😐  there are 20 strings to cover the existing combinations of addonType + query-or-not + category-or-not, for singular and plural.  With tags too it'll be an additional 20.

Having strings for each combination seems suboptimal to me but I could add them.

I considered we could ignore some less likely scenarios - tag pages only show with no category, and no query, so that could be the only scenario with a string mentioning the tag (addonType we'd always need as there's a filter for it) so 6 strings. But adding a `tag=xxx` parameter to a category or search results page then wouldn't mention the tag filter, and it would also apply to a tag page (e.g. `/tag/foo/?q=baa`) that has extra parameters added to the url - if the url is manually edited it's too easy to end up with an unsupported case that we wouldn't know what to do with, especially as `SearchContextCard` is used by search, category, and now tag pages.

Comment (eviljeff) [reply src/amo/components/SearchContextCard/index.js]: https://github.com/mozilla/addons-frontend/pull/10820/commits/c848cce9428199024e3755b16eed6342752e90b0 is what it looks like just to cover the cases we directly support with links and/or the UI (ignores combinations with categories or queries that are possible with specific url strings)

**Comment by willdurand** [comment]
> I've realised there's a limitation in just always dropping `tag` in the `SearchFilters` component - if there's a `tag=xxx` param in a vanilla search page you also lose it by changing the filters (e.g. change the sort) - not just actions like searching in the search box. Really, we want to retain `tag` when the UI filters are changed and only drop it when there's a new search.

mmm I don't think this is easily doable.

> We don't have this problem with searching with categories because we redirect search queries involving categories to the category pages. (Something we weren't planning to do with tag pages - and if we did we'd need a different url because of the nginx 301).

In this case, we should do what we've done for the categories. We can actually send a redirect because nginx does a 301 to `/tag/:tag`, which is different than our new URL pattern here (`/tag/:tag/` with the trailing slash).

File: src/amo/components/SearchContextCard/index.js [hunk 5]
```diff
@@ -83,6 +83,15 @@ export class SearchContextCardBase extends React.Component<InternalProps> {
               ),
               { count: i18n.formatNumber(count), query },
             );
+          } else if (tag) {
```
Comment (willdurand) [inline src/amo/components/SearchContextCard/index.js:86]: Will `addonType` be set somewhere?

Comment (eviljeff) [reply src/amo/components/SearchContextCard/index.js:86]: it can be set with the filters on the page

**Comment by willdurand** [comment]
>  The diff coverage is `94.11%`.

There seems to be missing coverage but I don't see anything obvious right now.

**Comment by willdurand** [comment]
>  The diff coverage is `94.11%`.

There seems to be missing coverage but I don't see anything obvious right now.

File: src/amo/components/SearchFilters/index.js [hunk 6]
```diff
@@ -74,8 +74,9 @@ export class SearchFiltersBase extends React.Component<InternalProps> {
       // If we are currently filtering by category, and the filter to change is 'sort',
       // force recommendations to the top.
       // See https://github.com/mozilla/addons-frontend/issues/8084
+      // We also do this for tag filtering.
```
Comment (willdurand) [inline src/amo/components/SearchFilters/index.js:77]:
```suggestion
      // We also do this for tag filtering because category and tag pages are very
      // similar.
```

**Comment by eviljeff** [comment]
> 
> 
> > The diff coverage is `94.11%`.
> 
> There seems to be missing coverage but I don't see anything obvious right now.

Oh, where do you see that?  I looked at https://app.codecov.io/gh/mozilla/addons-frontend/compare/10820/diff and read it as the overall code was 98.47% covered and it was 0.01% more covered after the patch (because more code with tests rather than more testing for existing code, I assumed)

File: tests/unit/amo/components/TestSearchFilters.js [hunk 7]
```diff
@@ -206,6 +206,33 @@ describe(__filename, () => {
...
+    select.simulate('change', createFakeEvent({ currentTarget }));
+
+    sinon.assert.calledWithExactly(fakeHistory.push, {
+      pathname: `/en-US/${CLIENT_APP_FIREFOX}/search/`,
+      query: convertFiltersToQueryParams({
+        query: 'Music player',
+        sort: `${SEARCH_SORT_RECOMMENDED},${SEARCH_SORT_TRENDING}`,
+      }),
+    });
+  });
```
[the remark this answers is not in the document]
Comment (eviljeff) [reply tests/unit/amo/components/TestSearchFilters.js:234]: I can duplicate the test... but I can't see what would be different in either test?  (Probably why the tests relating to categories are written in the same way)

**Comment by willdurand** [comment]
> Oh, where do you see that? I looked at [app.codecov.io/gh/mozilla/addons-frontend/compare/10820/diff](https://app.codecov.io/gh/mozilla/addons-frontend/compare/10820/diff) and read it as the overall code was 98.47% covered and it was 0.01% more covered after the patch (because more code with tests rather than more testing for existing code, I assumed)

In the codecov comment above. It says 100% now, though, so we're good!

File: tests/unit/amo/components/TestSearchFilters.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (willdurand) [reply tests/unit/amo/components/TestSearchFilters.js:234]: We shouldn't have to change both params (filters) at the same time, right?

[the remark this answers is not in the document]
Comment (eviljeff) [reply tests/unit/amo/components/TestSearchFilters.js:234]: The way I understood it, the trigger is the filter select being changed, not the params.  For the setup, `tag` needs to be specified to make this is a test that checks functionality around tags specifically, and sort isn't actually changed.  But I'll take code suggestions!

[the remark this answers is not in the document]
Comment (willdurand) [reply tests/unit/amo/components/TestSearchFilters.js:234]: ha well, nevermind, no need to waste too much time on this given that the test for categories looks the same 😁

**Review by willdurand — APPROVED** [review]
Looks good, thanks!
